pennylane/ops/op_math/controlled_class.py

```diff
@@ -0,0 +1,382 @@
+
+    # Methods ##########################################
+
+    def label(self, decimals=None, base_label=None, cache=None):
+        return self.base.label(decimals=decimals, base_label=base_label, cache=cache)
+
+    def matrix(self, wire_order=None):
+
+        base_matrix = self.base.matrix()
+        interface = qmlmath.get_interface(base_matrix)
+
+        num_target_states = 2 ** len(self.target_wires)
+        num_control_states = 2 ** len(self.control_wires)
+        total_matrix_size = num_control_states * num_target_states
+
+        padding_left = self._control_int * num_target_states
+        padding_right = total_matrix_size - padding_left - num_target_states
+
+        left_pad = qmlmath.cast_like(qmlmath.eye(padding_left, like=interface), 1j)
+        right_pad = qmlmath.cast_like(qmlmath.eye(padding_right, like=interface), 1j)
+
+        canonical_matrix = qmlmath.block_diag([left_pad, base_matrix, right_pad])
+
+        if wire_order is None or self.wires == Wires(wire_order):
+            return canonical_matrix
+
+        active_wires = self.control_wires + self.target_wires
+        return operation.expand_matrix(canonical_matrix, wires=active_wires, wire_order=wire_order)
+
+    # pylint: disable=arguments-differ
+    def sparse_matrix(self, wire_order=None, format="csr"):
+        if wire_order is not None:
+            raise NotImplementedError("wire_order argument is not yet implemented.")
+
+        try:
+            target_mat = self.base.sparse_matrix()
+        except operation.SparseMatrixUndefinedError:
+            try:
+                target_mat = sparse.lil_matrix(self.base.matrix())
+            except operation.MatrixUndefinedError as e:
+                raise operation.SparseMatrixUndefinedError from e
+
+        num_target_states = 2 ** len(self.target_wires)
+        num_control_states = 2 ** len(self.control_wires)
+        total_states = num_target_states * num_control_states
+
+        start_ind = self._control_int * num_target_states
+        end_ind = start_ind + num_target_states
+
+        m = sparse.eye(total_states, format="lil", dtype=target_mat.dtype)
+
+        m[start_ind:end_ind, start_ind:end_ind] = target_mat
+
+        return m.asformat(format=format)
+
+    def eigvals(self):
+        base_eigvals = self.base.eigvals()
+        num_target_wires = len(self.target_wires)
+        num_control_wires = len(self.control_wires)
+
+        total = 2 ** (num_target_wires + num_control_wires)
+        ones = np.ones(total - len(base_eigvals))
+
+        return qmlmath.concatenate([ones, base_eigvals])
+
+    def diagonalizing_gates(self):
+        return self.base.diagonalizing_gates()
+
+    def decomposition(self):
+        if not all(self.control_values):
+            d = [
+                qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val
+            ]
+            d += [Controlled(self.base, self.control_wires, work_wires=self.work_wires)]
```

Comment:
The decomposition is a complicated question.  Do we want to decompose in the number of control wires or in the target operation?

Since it's a tricky thing, I'm not addressing it in this PR.  Hopefully in successive PR's we can add the different algorithms and use them here.